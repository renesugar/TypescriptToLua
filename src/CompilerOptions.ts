import * as ts from "typescript";

export interface CompilerOptions extends ts.CompilerOptions {
    noHeader?: boolean;
    luaTarget?: LuaTarget;
    luaLibImport?: LuaLibImportKind;
    noHoisting?: boolean;
}

export enum LuaLibImportKind {
    None = "none",
    Always = "always",
    Inline = "inline",
    Require = "require",
}

export enum LuaTarget {
    Lua51 = "5.1",
    Lua52 = "5.2",
    Lua53 = "5.3",
    LuaJIT = "jit",
}
