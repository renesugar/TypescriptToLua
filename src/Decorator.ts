export class Decorator {
    public static isValid(decoratorKindString: string): boolean {
        return this.getDecoratorKind(decoratorKindString) !== undefined;
    }

    public static getDecoratorKind(decoratorKindString: string): DecoratorKind {
        switch (decoratorKindString.toLowerCase()) {
            case "extension":
                return DecoratorKind.Extension;
            case "metaextension":
                return DecoratorKind.MetaExtension;
            case "customconstructor":
                return DecoratorKind.CustomConstructor;
            case "compilemembersonly":
                return DecoratorKind.CompileMembersOnly;
            case "pureabstract":
                return DecoratorKind.PureAbstract;
            case "phantom":
                return DecoratorKind.Phantom;
            case "tuplereturn":
                return DecoratorKind.TupleReturn;
            case "noclassor":
                return DecoratorKind.NoClassOr;
            case "luaiterator":
                return DecoratorKind.LuaIterator;
        }

        return undefined;
    }

    public kind: DecoratorKind;
    public args: string[];

    constructor(name: string, args: string[]) {
        this.kind = Decorator.getDecoratorKind(name);
        this.args = args;
    }
}

export enum DecoratorKind {
    Extension = "Extension",
    MetaExtension = "MetaExtension",
    CustomConstructor = "CustomConstructor",
    CompileMembersOnly = "CompileMembersOnly",
    PureAbstract = "PureAbstract",
    Phantom = "Phantom",
    TupleReturn = "TupleReturn",
    NoClassOr = "NoClassOr",
    LuaIterator = "LuaIterator",
}
