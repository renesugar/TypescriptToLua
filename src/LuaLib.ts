import * as fs from "fs";
import * as path from "path";

export enum LuaLibFeature {
    ArrayConcat = "ArrayConcat",
    ArrayEvery = "ArrayEvery",
    ArrayFilter = "ArrayFilter",
    ArrayForEach = "ArrayForEach",
    ArrayFindIndex = "ArrayFindIndex",
    ArrayIndexOf = "ArrayIndexOf",
    ArrayMap = "ArrayMap",
    ArrayPush = "ArrayPush",
    ArrayReverse = "ArrayReverse",
    ArrayShift = "ArrayShift",
    ArrayUnshift = "ArrayUnshift",
    ArraySort = "ArraySort",
    ArraySlice = "ArraySlice",
    ArraySome = "ArraySome",
    ArraySplice = "ArraySplice",
    ClassIndex = "ClassIndex",
    ClassNewIndex = "ClassNewIndex",
    FunctionApply = "FunctionApply",
    FunctionBind = "FunctionBind",
    FunctionCall = "FunctionCall",
    Index = "Index",
    InstanceOf = "InstanceOf",
    Iterator = "Iterator",
    Map = "Map",
    NewIndex = "NewIndex",
    ObjectAssign = "ObjectAssign",
    ObjectEntries = "ObjectEntries",
    ObjectKeys = "ObjectKeys",
    ObjectValues = "ObjectValues",
    Set = "Set",
    WeakMap = "WeakMap",
    WeakSet = "WeakSet",
    StringReplace = "StringReplace",
    StringSplit = "StringSplit",
    StringConcat = "StringConcat",
    Symbol = "Symbol",
    SymbolRegistry = "SymbolRegistry",
}

const luaLibDependencies: {[lib in LuaLibFeature]?: LuaLibFeature[]} = {
    Iterator: [LuaLibFeature.Symbol],
    Map: [LuaLibFeature.InstanceOf, LuaLibFeature.Iterator, LuaLibFeature.Symbol],
    Set: [LuaLibFeature.InstanceOf, LuaLibFeature.Iterator, LuaLibFeature.Symbol],
    WeakMap: [LuaLibFeature.InstanceOf, LuaLibFeature.Iterator, LuaLibFeature.Symbol],
    WeakSet: [LuaLibFeature.InstanceOf, LuaLibFeature.Iterator, LuaLibFeature.Symbol],
    SymbolRegistry: [LuaLibFeature.Symbol],
};

export class LuaLib {
    public static loadFeatures(features: Iterable<LuaLibFeature>): string {
        let result = "";

        const loadedFeatures = new Set<LuaLibFeature>();

        function load(feature: LuaLibFeature): void {
            if (!loadedFeatures.has(feature)) {
                loadedFeatures.add(feature);
                if (luaLibDependencies[feature]) {
                    luaLibDependencies[feature].forEach(load);
                }
                const featureFile = path.resolve(__dirname, `../dist/lualib/${feature}.lua`);
                result += fs.readFileSync(featureFile).toString() + "\n";
            }
        }

        for (const feature of features) {
            load(feature);
        }
        return result;
    }
}
