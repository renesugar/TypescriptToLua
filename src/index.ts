export {parseCommandLine} from "./CommandLineParser";
export {compile, compileFilesWithOptions, transpileString, watchWithOptions} from "./Compiler";
export {CompilerOptions, LuaLibImportKind, LuaTarget,} from "./CompilerOptions";
export {LuaLibFeature,} from "./LuaLib";
export {LuaTranspiler,} from "./LuaTranspiler";
