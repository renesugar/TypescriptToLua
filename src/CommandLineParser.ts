import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";

import {CompilerOptions, LuaTarget, LuaLibImportKind} from "./CompilerOptions";

export type CLIParseResult = ParseResult<ParsedCommandLine>;

type ParseResult<T> =
    { isValid: true; result: T }
    | { isValid: false, errorMessage: string};

type ArgumentParseResult<T> =
    { isValid: true; result: T; increment?: number }
    | { isValid: false, errorMessage: string };

interface ParsedCommandLine extends ts.ParsedCommandLine {
    options: CompilerOptions;
}

interface BaseCLIOption {
    aliases: string[];
    describe: string;
    type: string;
}

interface CLIOption<T> extends BaseCLIOption {
    choices: T[];
    default: T;
}

const optionDeclarations: {[key: string]: CLIOption<any>} = {
    luaLibImport: {
        choices: [LuaLibImportKind.Inline, LuaLibImportKind.Require, LuaLibImportKind.Always, LuaLibImportKind.None],
        default: LuaLibImportKind.Inline,
        describe: "Specifies how js standard features missing in lua are imported.",
        type: "enum",
    } as CLIOption<string>,
    luaTarget: {
        aliases: ["lt"],
        choices: [LuaTarget.LuaJIT, LuaTarget.Lua53, LuaTarget.Lua52, LuaTarget.Lua51],
        default: LuaTarget.LuaJIT,
        describe: "Specify Lua target version.",
        type: "enum",
    } as CLIOption<string>,
    noHeader: {
        default: false,
        describe: "Specify if a header will be added to compiled files.",
        type: "boolean",
    } as CLIOption<boolean>,
    noHoisting: {
        default: false,
        describe: "Disables hoisting.",
        type: "boolean",
    } as CLIOption<boolean>,
};

export const { version } = require("../package.json");

const helpString =
    `Version ${version}\n` +
    "Syntax:   tstl [options] [files...]\n\n" +

    "Examples: tstl path/to/file.ts [...]\n" +
    "          tstl -p path/to/tsconfig.json\n\n" +

    "In addition to the options listed below you can also pass options\n" +
    "for the typescript compiler (For a list of options use tsc -h).\n" +
    "Some tsc options might have no effect.";

/**
 * Parse the supplied arguments.
 * The result will include arguments supplied via CLI and arguments from tsconfig.
 */
export function parseCommandLine(args: string[]): CLIParseResult
{
    let commandLine = ts.parseCommandLine(args);

    // Run diagnostics to check for invalid tsc options
    const diagnosticsResult = runTsDiagnostics(commandLine);
    if (diagnosticsResult.isValid === false) {
        return diagnosticsResult;
    }

    // This will add TS and TSTL options from a tsconfig
    const configResult = readTsConfig(commandLine);
    if (configResult.isValid === true) {
        commandLine = configResult.result;
    } else {
        return { isValid: false, errorMessage: configResult.errorMessage };
    }

    // Run diagnostics to check for invalid tsconfig
    const diagnosticsResult2 = runTsDiagnostics(commandLine);
    if (diagnosticsResult2.isValid === false) {
        return diagnosticsResult2;
    }

    // Merge TSTL CLI options in (highest priority) will also set defaults if none specified
    const tstlCLIResult = parseTSTLOptions(commandLine, args);
    if (tstlCLIResult.isValid === true) {
        commandLine = tstlCLIResult.result;
    } else {
        return { isValid: false, errorMessage: tstlCLIResult.errorMessage };
    }

    if (commandLine.options.project && !commandLine.options.rootDir) {
        commandLine.options.rootDir = path.dirname(commandLine.options.project);
    }

    if (!commandLine.options.rootDir) {
        commandLine.options.rootDir = process.cwd();
    }

    if (!commandLine.options.outDir) {
        commandLine.options.outDir = commandLine.options.rootDir;
    }

    return { isValid: true, result: commandLine as ParsedCommandLine };
}

export function getHelpString(): string {
    let result = helpString + "\n\n";

    result += "Options:\n";
    for (const optionName in optionDeclarations) {
        const option = optionDeclarations[optionName];
        const aliasStrings = option.aliases
            ? option.aliases.map(a => "-" + a)
            : [];

        const optionString = aliasStrings.concat(["--" + optionName]).join("|");

        const parameterDescribe = option.choices
            ? option.choices.join("|")
            : option.type;

        const spacing = " ".repeat(Math.max(1, 45 - optionString.length - parameterDescribe.length));

        result += `\n ${optionString} <${parameterDescribe}>${spacing}${option.describe}\n`;
    }

    return result;
}

function readTsConfig(parsedCommandLine: ts.ParsedCommandLine): CLIParseResult
{
    const options = parsedCommandLine.options;

    // Load config
    if (options.project) {
        const findProjectPathResult = findConfigFile(options);
        if (findProjectPathResult.isValid === true) {
            options.project = findProjectPathResult.result;
        } else {
            return { isValid: false, errorMessage: findProjectPathResult.errorMessage };
        }

        const configPath = options.project;
        const configContents = fs.readFileSync(configPath).toString();
        const configJson = ts.parseConfigFileTextToJson(configPath, configContents);
        const parsedJsonConfig = ts.parseJsonConfigFileContent(
            configJson.config,
            ts.sys,
            path.dirname(configPath),
            options
        );

        for (const key in parsedJsonConfig.raw) {
            const option = optionDeclarations[key];
            if (option !== undefined) {
                const value = readValue(parsedJsonConfig.raw[key], option.type, key);
                if (option.choices) {
                    if (option.choices.indexOf(value) < 0) {
                        return {
                            isValid: false,
                            errorMessage: `Unknown ${key} value '${value}'.\nAccepted values: ${option.choices}`,
                        };
                    }
                }

                parsedJsonConfig.options[key] = value;
            }
        }
        return { isValid: true, result: parsedJsonConfig };
    }
    return { isValid: true, result: parsedCommandLine };
}

function parseTSTLOptions(commandLine: ts.ParsedCommandLine, args: string[]): CLIParseResult {
    const result = {};
    for (let i = 0; i < args.length; i++) {
        if (args[i].startsWith("--")) {
            const argumentName = args[i].substr(2);
            const option = optionDeclarations[argumentName];
            if (option) {
                const argumentResult = getArgumentValue(argumentName, i, args);
                if (argumentResult.isValid === true) {
                    result[argumentName] = argumentResult.result;
                    // Skip value from being considered as option
                    i += argumentResult.increment !== undefined ? argumentResult.increment : 1;
                } else {
                    return { isValid: false, errorMessage: argumentResult.errorMessage };
                }
            }
        } else if (args[i].startsWith("-")) {
            const argument = args[i].substr(1);
            let argumentName: string;
            for (const key in optionDeclarations) {
                if (optionDeclarations[key].aliases && optionDeclarations[key].aliases.indexOf(argument) >= 0) {
                    argumentName = key;
                    break;
                }
            }

            if (argumentName) {
                const argumentResult = getArgumentValue(argumentName, i, args);
                if (argumentResult.isValid === true) {
                    result[argumentName] = argumentResult.result;
                    // Skip value from being considered as option
                    i += argumentResult.increment !== undefined ? argumentResult.increment : 1;
                } else {
                    return { isValid: false, errorMessage: argumentResult.errorMessage };
                }
            }
        }
    }
    for (const option in result) {
        commandLine.options[option] = result[option];
    }
    // Add defaults if not set
    const defaultOptions = getDefaultOptions();
    for (const option in defaultOptions) {
        if (!commandLine.options[option]) {
            commandLine.options[option] = defaultOptions[option];
        }
    }
    return { isValid: true, result: commandLine };
}

function getArgumentValue(
    argumentName: string,
    argumentIndex: number,
    args: string[]
): ArgumentParseResult<string | boolean>
{
    const option = optionDeclarations[argumentName];
    const argument = args[argumentIndex + 1];

    if (option.type === "boolean" && (argument === undefined || argument.startsWith("-"))) {
        // Set boolean arguments without supplied value to true
        return { isValid: true, result: true, increment: 0 };
    }

    if (argument === undefined) {
        return { isValid: false, errorMessage: `Missing value for parameter ${argumentName}`};
    }

    const value = readValue(argument, option.type, argumentName);

    if (option.choices) {
        if (option.choices.indexOf(value) < 0) {
            return {
                isValid: false,
                errorMessage: `Unknown ${argumentName} value '${value}'. Accepted values are: ${option.choices}`,
            };
        }
    }

    return { isValid: true, result: value };
}

function readValue(valueString: string, valueType: string, parameterName: string): string | boolean {
    if (valueType === "boolean") {
        return valueString === "true" || valueString === "t"
            ? true
            : false;
    } else if (valueType === "enum") {
        return valueString.toLowerCase();
    } else {
        return valueString;
    }
}

function getDefaultOptions(): CompilerOptions {
    const options: CompilerOptions = {};

    for (const optionName in optionDeclarations) {
        if (optionDeclarations[optionName].default !== undefined) {
            options[optionName] = optionDeclarations[optionName].default;
        }
    }

    return options;
}

/** Check the current state of the ParsedCommandLine for errors */
function runTsDiagnostics(commandLine: ts.ParsedCommandLine): ParseResult<boolean> {
    // Remove files that dont exist
    commandLine.fileNames = commandLine.fileNames.filter(file => fs.existsSync(file) || fs.existsSync(file + ".ts"));

    const tsInvalidCompilerOptionErrorCode = 5023;
    if (commandLine.errors.length !== 0) {
        // Generate a list of valid option names and aliases
        const optionNames: string[] = [];
        for (const key of Object.keys(optionDeclarations)) {
            optionNames.push(key);
            const alias = optionDeclarations[key].aliases;
            if (alias) {
                if (typeof alias === "string") {
                    optionNames.push(alias);
                } else {
                    optionNames.push(...alias);
                }
            }
        }

        for (const err of commandLine.errors) {
            // Ignore errors caused by tstl specific compiler options
            if (err.code === tsInvalidCompilerOptionErrorCode) {
                let ignore = false;
                for (const optionName of optionNames) {
                    if (err.messageText.toString().indexOf(optionName) !== -1) {
                        ignore = true;
                        break;
                    }
                }

                if (!ignore) {
                    return { isValid: false, errorMessage: `error TS${err.code}: ${err.messageText}`};
                }
            }
        }
    }

    return { isValid: true, result: true };
}

/** Find configFile, function from ts api seems to be broken? */
export function findConfigFile(options: ts.CompilerOptions): ParseResult<string> {
    if (!options.project) {
        return { isValid: false, errorMessage: `error no base path provided, could not find config.`};
    }
    let configPath = options.project;
    // If the project path is wrapped in double quotes, remove them
    if (/^".*"$/.test(configPath)) {
        configPath = configPath.substring(1, configPath.length - 1);
    }
    /* istanbul ignore if: Testing else part is not really possible via automated tests */
    if (!path.isAbsolute(configPath)) {
        // TODO check if options.project can even contain non absolute paths
        configPath = path.join(process.cwd(), configPath);
    }
    if (fs.statSync(configPath).isDirectory()) {
        configPath = path.join(configPath, "tsconfig.json");
    } else if (fs.statSync(configPath).isFile() && path.extname(configPath) === ".ts") {
        // Search for tsconfig upwards in directory hierarchy starting from the file path
        const dir = path.dirname(configPath).split(path.sep);
        for (let i = dir.length; i > 0; i--) {
            const searchPath = dir.slice(0, i).join("/") + path.sep + "tsconfig.json";

            // If tsconfig.json was found, stop searching
            if (ts.sys.fileExists(searchPath)) {
                configPath = searchPath;
                break;
            }
        }
    }

    return { isValid: true, result: configPath };
}
