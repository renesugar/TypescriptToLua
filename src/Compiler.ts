import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";

import * as CommandLineParser from "./CommandLineParser";
import {CompilerOptions, LuaLibImportKind, LuaTarget} from "./CompilerOptions";
import {LuaTranspiler} from "./LuaTranspiler";

export function compile(argv: string[]): void {
    const parseResult = CommandLineParser.parseCommandLine(argv);

    if (parseResult.isValid === true) {

        if (parseResult.result.options.help) {
            console.log(CommandLineParser.getHelpString());
            return;
        }

        if (parseResult.result.options.version) {
            console.log(CommandLineParser.version);
            return;
        }

        /* istanbul ignore if: tested in test/compiler/watchmode.spec with subproccess */
        if (parseResult.result.options.watch) {
            watchWithOptions(parseResult.result.fileNames, parseResult.result.options);
        } else {
            compileFilesWithOptions(parseResult.result.fileNames, parseResult.result.options);
        }
    } else {
        console.error(`Invalid CLI input: ${parseResult.errorMessage}`);
    }
}

/* istanbul ignore next: tested in test/compiler/watchmode.spec with subproccess */
export function watchWithOptions(fileNames: string[], options: CompilerOptions): void {
    let host: ts.WatchCompilerHost<ts.SemanticDiagnosticsBuilderProgram>;
    let config = false;
    if (options.project) {
        config = true;
        host = ts.createWatchCompilerHost(options.project, options, ts.sys, ts.createSemanticDiagnosticsBuilderProgram);
    } else {
        host = ts.createWatchCompilerHost(fileNames, options, ts.sys, ts.createSemanticDiagnosticsBuilderProgram);
    }

    host.afterProgramCreate = program => {
        const transpiler = new LuaTranspiler(program.getProgram());

        const status = transpiler.emitFilesAndReportErrors();
        const errorDiagnostic: ts.Diagnostic = {
            category: undefined,
            code: 6194,
            file: undefined,
            length: 0,
            messageText: "Found 0 errors. Watching for file changes.",
            start: 0,
        };
        if (status !== 0) {
            errorDiagnostic.messageText = "Found Errors. Watching for file changes.";
            errorDiagnostic.code = 6193;
        }
        host.onWatchStatusChange(errorDiagnostic, host.getNewLine(), program.getCompilerOptions());
    };

    if (config) {
        ts.createWatchProgram(
            host as ts.WatchCompilerHostOfConfigFile<ts.SemanticDiagnosticsBuilderProgram>
        );
    } else {
        ts.createWatchProgram(
            host as ts.WatchCompilerHostOfFilesAndCompilerOptions<ts.SemanticDiagnosticsBuilderProgram>
        );
    }
}

export function compileFilesWithOptions(fileNames: string[], options: CompilerOptions): void {
    const program = ts.createProgram(fileNames, options);

    const transpiler = new LuaTranspiler(program);

    transpiler.emitFilesAndReportErrors();
}

const libCache: {[key: string]: string} = {};

const defaultCompilerOptions: CompilerOptions = {
    luaLibImport: LuaLibImportKind.Require,
    luaTarget: LuaTarget.Lua53,
};

export function createStringCompilerProgram(
    input: string, options: CompilerOptions = defaultCompilerOptions, filePath = "file.ts"): ts.Program {
    const compilerHost =  {
        directoryExists: () => true,
        fileExists: (fileName): boolean => true,
        getCanonicalFileName: fileName => fileName,
        getCurrentDirectory: () => "",
        getDefaultLibFileName: () => "lib.es6.d.ts",
        getDirectories: () => [],
        getNewLine: () => "\n",

        getSourceFile: (filename: string) => {
            if (filename === filePath) {
                return ts.createSourceFile(filename, input, ts.ScriptTarget.Latest, false);
            }
            if (filename.indexOf(".d.ts") !== -1)Â {
                if (!libCache[filename]) {
                    const typeScriptDir = path.dirname(require.resolve("typescript"));
                    const filePath = path.join(typeScriptDir, filename);
                    if (fs.existsSync(filePath)) {
                        libCache[filename] = fs.readFileSync(filePath).toString();
                    } else {
                        const pathWithLibPrefix = path.join(typeScriptDir, "lib." + filename);
                        libCache[filename] = fs.readFileSync(pathWithLibPrefix).toString();
                    }
                }
                return ts.createSourceFile(filename, libCache[filename], ts.ScriptTarget.Latest, false);
            }
            return undefined;
        },

        readFile: () => "",

        useCaseSensitiveFileNames: () => false,
        // Don't write output
        writeFile: (name, text, writeByteOrderMark) => undefined,
    };
    return ts.createProgram([filePath], options, compilerHost);
}

export function transpileString(
    str: string,
    options: CompilerOptions = defaultCompilerOptions,
    ignoreDiagnostics = false,
    filePath = "file.ts"
): string {
    const program = createStringCompilerProgram(str, options, filePath);

    if (!ignoreDiagnostics) {
        const diagnostics = ts.getPreEmitDiagnostics(program);
        const typeScriptErrors = diagnostics.filter(diag => diag.category === ts.DiagnosticCategory.Error);

        if (typeScriptErrors.length > 0) {
            typeScriptErrors.forEach(e => console.warn(e.messageText));
            throw new Error("Encountered invalid TypeScript.");
        }
    }

    const transpiler = new LuaTranspiler(program);

    const result = transpiler.transpileSourceFile(program.getSourceFile(filePath));

    return result.trim();
}
