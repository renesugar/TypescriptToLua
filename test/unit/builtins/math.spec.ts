import * as tstl from "../../../src";
import * as util from "../../util";

test.each([
    "Math.cos()",
    "Math.sin()",
    "Math.min()",
    "Math.log2(3)",
    "Math.log10(3)",
    "const x = Math.log2(3)",
    "const x = Math.log10(3)",
    "Math.log1p(3)",
    "Math.round(3.3)",
    "Math.PI",
])("%s", code => {
    // TODO: Remove?
    util.testFunction(code).disableSemanticCheck().expectLuaToMatchSnapshot();
});

test.each(["E", "LN10", "LN2", "LOG10E", "LOG2E", "SQRT1_2", "SQRT2"])("Math.%s", constant => {
    util.testExpression`Math.${constant}`.tap(builder => {
        expect(builder.getLuaExecutionResult()).toBeCloseTo(builder.getJsExecutionResult());
    });
});

const expectMathAtan2: util.TapCallback = builder => expect(builder.getMainLuaCodeChunk()).toContain("math.atan2(");
const expectMathAtan: util.TapCallback = builder => expect(builder.getMainLuaCodeChunk()).toContain("math.atan(");
const expectLualibMathAtan2: util.TapCallback = builder =>
    expect(builder.getMainLuaCodeChunk()).toContain("__TS__MathAtan2(");

util.testEachVersion("Math.atan2", () => util.testExpression`Math.atan2(4, 5)`, {
    [tstl.LuaTarget.Universal]: builder => builder.tap(expectLualibMathAtan2),
    [tstl.LuaTarget.LuaJIT]: builder => builder.tap(expectMathAtan2),
    [tstl.LuaTarget.Lua51]: builder => builder.tap(expectMathAtan2),
    [tstl.LuaTarget.Lua52]: builder => builder.tap(expectMathAtan2),
    [tstl.LuaTarget.Lua53]: builder => builder.tap(expectMathAtan),
});

test("Math.atan2(4, 5)", () => {
    util.testExpression`Math.atan2(4, 5)`.expectToMatchJsResult();
});
