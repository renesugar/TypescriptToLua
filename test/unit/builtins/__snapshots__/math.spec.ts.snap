// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Math.PI 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local ____ = math.pi
end
return ____exports"
`;

exports[`Math.cos() 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    math.cos()
end
return ____exports"
`;

exports[`Math.log1p(3) 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    math.log(1 + 3)
end
return ____exports"
`;

exports[`Math.log2(3) 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local ____ = math.log(3) / 0.6931471805599453
end
return ____exports"
`;

exports[`Math.log10(3) 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local ____ = math.log(3) / 2.302585092994046
end
return ____exports"
`;

exports[`Math.min() 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    math.min()
end
return ____exports"
`;

exports[`Math.round(3.3) 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    math.floor(3.3 + 0.5)
end
return ____exports"
`;

exports[`Math.sin() 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    math.sin()
end
return ____exports"
`;

exports[`const x = Math.log2(3) 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local x = math.log(3) / 0.6931471805599453
end
return ____exports"
`;

exports[`const x = Math.log10(3) 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local x = math.log(3) / 2.302585092994046
end
return ____exports"
`;
